plutus-ledger/src/Ledger/Tx.hs

```diff
@@ -75,99 +73,119 @@ import Data.Proxy (Proxy (Proxy))
 import Data.Set (Set)
 import Data.Set qualified as Set
 import GHC.Generics (Generic)
-import Ledger.Address (Address, PaymentPubKey, StakePubKey, pubKeyAddress)
+import Ledger.Address (PaymentPubKey, StakePubKey, pubKeyAddress)
 import Ledger.Crypto (Passphrase, signTx, signTx', toPublicKey)
 import Ledger.Orphans ()
 import Ledger.Slot (SlotRange)
 import Ledger.Tx.CardanoAPI (SomeCardanoApiTx (SomeTx), ToCardanoError (..))
 import Ledger.Tx.CardanoAPI qualified as CardanoAPI
 import Ledger.Tx.Internal hiding (updateUtxoCollateral)
-import Plutus.Script.Utils.V1.Scripts (datumHash)
+import Plutus.Script.Utils.V1.Scripts (datumHash, validatorHash)
 import Plutus.V1.Ledger.Api qualified as V1
-import Plutus.V2.Ledger.Api qualified as V2
 -- for re-export
+import Plutus.V1.Ledger.Address qualified as V1
 import Plutus.V1.Ledger.Tx
 import Plutus.V1.Ledger.Tx qualified as V1.Tx
 import Prettyprinter (Pretty (pretty), braces, colon, hang, nest, viaShow, vsep, (<+>))
 
 type PrivateKey = Crypto.XPrv
 
--- | Transaction output that comes from a chain index query.
---
--- It is defined here instead of the plutus-chain-index because plutus-ledger
--- uses that datatype, and plutus-ledger can't depend on plutus-chain-index
--- because of a cyclic dependency.
---
--- This datatype was created in order to be used in
--- 'Ledger.Constraints.processConstraint', specifically with the constraints
--- 'MustSpendPubKeyOutput' and 'MustSpendScriptOutput'.
-data ChainIndexTxOut =
-    PublicKeyChainIndexTxOut {
-      -- | Address of the transaction output. The address is protected by a
-      -- public key hash.
-      _ciTxOutAddress         :: Address,
+-- | Transaction output represented off-chain
+-- add more
+data OffChainTxOut =
+    PublicKeyOffChainTxOut' {
+      -- | Hash of the public key protecting the transaction output.
+      ocTxOutAddress            :: V1.PubKeyHash,
```

Comment:
`V1.Address` instead of `V1.PubKeyHash`

Reply:
I actually meant to use `PubKeyHash` here, it's the name that's wrong. `ocTxOutAddress` will be an accessor function `ocTxOutAddress :: OffChainTxOut -> Address`.

Reply:
This is because `PubKeyHash` is actual evidence the txOut is protected by a public key, while `Address` can stil be a script.

Reply:
The problem is that were losing stake information. I don't see the problem of simply using `Address` here instead of requiring users to reconstruct the `Address` with the PubKeyHash and stake information everytime we want to get the address. Users would need to use `Address` instead of `PubKeyHash`

Reply:
> The problem is that were losing stake information

That can be added back.

> I don't see the problem of simply using Address here instead of requiring users to reconstruct the Address with the PubKeyHash and stake information everytime we want to get the address.

The goal was to make the following invalid state unrepresentable:

```
PublicKeyChainIndexTxOut (scriptHashAddress "abcd") mempty
```

Reply:
Ah I see. Then we just need to add staking information as a new field?

Reply:
Yup, we can do that.